ui/src/app/components/admin/audit-page-component.tsx

```diff
@@ -0,0 +1,145 @@
+   */
+  query: string;
+  /**
+   * ID in the MySQL database and BigQuery Audit Database for the this qyery. Currently either
+   * a userId or workspaceId as appropriate. This may be needed for situations (such as workspacd audit)
+   * where it's not obvious in the responses which workspace is the one you audited.
+   */
+  sourceId: number;
+}
+
+// Common properties for User & Workspace (and similar future pages).
+// Assmptions: the path parameter is called initially with the audit API subject.
+//   (this will need revisiting for multi-subject queries)
+export interface AuditPageProps {
+  initialAuditSubject?: string;
+  auditSubjectType: string;
+  queryAuditLog: (subject: string) => Promise<GenericAuditQueryResult>;
+  getNextAuditPath: (subject: string) => string;
+  debug: boolean;
+  buttonLabel?: string;
+}
+
+const UserInput = ({initialAuditSubject, auditSubjectType, getNextAuditPath, buttonLabel}) => {
+  const [auditSubject, setAuditSubject] = useState(initialAuditSubject);
+  const [loadNextSubject, setLoadNextSubject] = useState(false);
+  useEffect(() => {
+    loadNextSubject && setLoadNextSubject(false);
+  }, [loadNextSubject]);
+
+  return <React.Fragment>
+    {loadNextSubject && <Navigate to={getNextAuditPath(auditSubject)}/>}
+    <TextInputWithLabel
+      containerStyle={{display: 'inline-block'}}
+      style={{width: '15rem', margin: '1rem'}}
+      labelText = {buttonLabel || auditSubjectType}
+      value = {auditSubject}
+      onChange = {setAuditSubject}
+    />
+    <Button style={{height: '1.5rem', margin: '0.25rem 0.5rem'}} disabled={fp.isEmpty(auditSubject)} onClick={() => setLoadNextSubject(true)}>
+    Audit
+    </Button>
+  </React.Fragment>;
+};
+
+const NumActions = ({onChange, totalActions}) => {
+  const [displayNum, setDisplayNum] = useState(20);
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => onChange(displayNum), 250);
+    return () => clearTimeout(timeoutId);
+  }, [displayNum]);
+
+  return <div>
+    <label style={{
+      fontSize: 14,
+      display: 'block',
+      color: colors.primary,
+      lineHeight: '22px',
+      fontWeight: 600,
+      marginRight: '0.25rem'
+    }}>{`Number of Actions to Display (${totalActions} available)`}</label>
+    <NumberInput value={Math.min(displayNum, totalActions)} min={1} max={totalActions} style={{width: '4rem'}} onChange={setDisplayNum}/>
+  </div>;
+};
+
+export const AuditPageComponent = (props: AuditPageProps) => {
+  const {initialAuditSubject, queryAuditLog, getNextAuditPath, debug, auditSubjectType, buttonLabel} = props;
+  const emptyResult = {actions: [], logEntries: [], sourceId: 0, query: ''};
+  const [loading, setLoading] = useState(true);
+  const [queryResult, setQueryResult] = useState<GenericAuditQueryResult>(emptyResult);
+  const [displayNum, setDisplayNum] = useState(20);
+  const {actions, sourceId, query} = queryResult;
+
+  useEffect(() => {
+    if (debug) {
```

Comment:
extraneous?

Reply:
I'd like to have something like Log Levels that come with the Java loggers, but it looks like I don't get that for free. Basically I want it to be an easier experience doing the next api type.

Maybe changing `debug` to `verboseLogging` would be more respectable?

Reply:
I guess what I don't understand is why you'd console-log information that could be displayed on the screen.

Reply:
I agree with Ariel. When developing it wouldn't/shouldn't be difficult to add a `console.log` of query results if needed. I am not a fan of leaving debug statements in, even if hidden behind a flag. 

One reason for this is these statements can start leaking into the console unintentionally if we allow `console` statements in the code base at all. Having a lot of `console` messages can add noise to a search and the code. Unfortunately, right now there are a bunch of console statements for errors in the code rather than error notifications that are useful to the user (something showing up in the UI). I would prefer to not add more.

Personally, I would like to see a lint rule added in the not too distant future prohibiting the use of `console.log`. I believe that this helps keep the UI code clean and forces good (at least better) error notifications for users.

Reply:
I hear ya. Without some kind of a log level we can adjust it's hard to have conditional output. I wouldn't kill all logging until there was a legitimate way to have optional diagnostic statements that end users couldn't turn on.

Can we have an unobtrusive widget on the page to show it? Or even just a button launching a modal? Call it `[Show BigQuery SQL]` or something.

Reply:
Yes, I'd prefer to have it on the screen, behind a fold (e.g. the admin or profile dropdowns in the sidebar). I would personally find a modal annoying.